src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/FormattingHelpers.cs

```diff
@@ -369,74 +372,84 @@ public static bool IsNestedQueryExpression(this SyntaxToken token)
             {
-                return true;
+                return anonymousObjectInitializer.Parent is AnonymousObjectCreationExpressionSyntax;
             }
+            else
+            {
+                return false;
+            }
+        }
 
-            if (parent is ObjectCreationExpressionSyntax)
+        public static bool IsInitializerForArrayOrCollectionCreationExpression(this SyntaxNode node)
+        {
+            if (node is InitializerExpressionSyntax initializer)
             {
-                if (initializer.Expressions.Count <= 0)
+                var parent = initializer.Parent;
+                if (parent is ArrayCreationExpressionSyntax ||
+                    parent is ImplicitArrayCreationExpressionSyntax ||
+                    parent is EqualsValueClauseSyntax ||
+                    parent.IsKind(SyntaxKind.SimpleAssignmentExpression))
                 {
                     return true;
                 }
 
-                var expression = initializer.Expressions[0];
-                if (expression.Kind() == SyntaxKind.SimpleAssignmentExpression)
+                if (parent is ObjectCreationExpressionSyntax)
                 {
-                    return true;
+                    return !IsInitializerForObjectOrAnonymousObjectCreationExpression(initializer);
                 }
-            }
 
-            return false;
-        }
-
-        public static bool IsInitializerForArrayOrCollectionCreationExpression(this SyntaxNode node)
-        {
-            var initializer = node as InitializerExpressionSyntax;
-            AnonymousObjectMemberDeclaratorSyntax anonymousObjectInitializer = null;
-            if (initializer == null)
-            {
-                anonymousObjectInitializer = node as AnonymousObjectMemberDeclaratorSyntax;
-                if (anonymousObjectInitializer == null)
+                return false;
+            }
+            else if (node is AnonymousObjectMemberDeclaratorSyntax anonymousObjectInitializer)
+            {
+                var parent = anonymousObjectInitializer.Parent;
+                if (parent is ArrayCreationExpressionSyntax ||
+                    parent is ImplicitArrayCreationExpressionSyntax ||
+                    parent is EqualsValueClauseSyntax ||
+                    parent is ObjectCreationExpressionSyntax ||
+                    parent.IsKind(SyntaxKind.SimpleAssignmentExpression))
                 {
-                    return false;
+                    return true;
                 }
-            }
 
-            var parent = initializer != null ? initializer.Parent : anonymousObjectInitializer.Parent;
-            if (parent is ArrayCreationExpressionSyntax ||
-                parent is ImplicitArrayCreationExpressionSyntax ||
-                parent is EqualsValueClauseSyntax ||
-                parent.Kind() == SyntaxKind.SimpleAssignmentExpression)
-            {
-                return true;
+                return false;
```

Comment:
nit: considering just making this a single return of the above condition.

Reply:
I'm OK with this for now. The branching is weird in order to fully leverage nullable reference types flow analysis but visually consistent.